### ISSUE TITLE
False positive with rubygem activesupport (rails) 

### ISSUE THREAD

**Issue by tspascoal** [aquasecurity/trivy#702]
**Description**

**What did you expect to happen?**

No vulnerabilities reported

**What happened instead?**

Reporting a vulnerability  ([CVE-2020-8165](https://nvd.nist.gov/vuln/detail/CVE-2020-8165) which states only < 6.0.3.1 is vulnerable) on a version that no longer is vulnerable

Version being detected is `6.0.3.4` whereas the fix was made in `activesupport 6.0.3.1`

If I use a dummy version of 6.0.4.1 then no vulnerabilities are found

**Output of run with `-debug`:**

```
2020-10-19T02:19:09.109+0100    DEBUG   Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2020-10-19T02:19:09.117+0100    DEBUG   cache dir:  /home/tspascoal/.cache/trivy
2020-10-19T02:19:09.117+0100    DEBUG   DB update was skipped because DB is the latest
2020-10-19T02:19:09.117+0100    DEBUG   DB Schema: 1, Type: 1, UpdatedAt: 2020-10-19 12:19:49.968147945 +0000 UTC, NextUpdate: 2020-10-20 00:19:49.968147445 +0000 UTC
2020-10-19T02:19:09.117+0100    DEBUG   Vulnerability type:  [os library]
2020-10-19T02:19:09.132+0100    DEBUG   Artifact ID: sha256:d05810661b660e9dde2a52ab1231a7c64d441fab50725bb4cceb39b4af9a1690
2020-10-19T02:19:09.133+0100    DEBUG   Blob IDs: [sha256:d05810661b660e9dde2a52ab1231a7c64d441fab50725bb4cceb39b4af9a1690]
2020-10-19T02:19:09.133+0100    WARN    OS is not detected and vulnerabilities in OS packages are not detected.
2020-10-19T02:19:09.133+0100    DEBUG   Detecting library vulnerabilities, path: Gemfile.lock
2020-10-19T02:19:09.133+0100    INFO    Detecting ruby vulnerabilities...

Gemfile.lock
============
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 1)

+---------------+------------------+----------+-------------------+------------------+--------------------------------+
|    LIBRARY    | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |  FIXED VERSION   |             TITLE              |
+---------------+------------------+----------+-------------------+------------------+--------------------------------+
| activesupport | CVE-2020-8165    | CRITICAL | 6.0.3-4           | 6.0.3.1, 5.2.4.3 | rubygem-activesupport:         |
|               |                  |          |                   |                  | potentially unintended         |
|               |                  |          |                   |                  | unmarshalling of user-provided |
|               |                  |          |                   |                  | objects in MemCacheStore and   |
|               |                  |          |                   |                  | RedisCacheStore                |
+---------------+------------------+----------+-------------------+------------------+-------------------------------
```

**Output of `trivy -v`:**

```
2020-10-19T02:19:09.109+0100    DEBUG   Severities: UNKNOWN,LOW,MEDIUM,HIGH,CRITICAL
2020-10-19T02:19:09.117+0100    DEBUG   cache dir:  /home/tspascoal/.cache/trivy
2020-10-19T02:19:09.117+0100    DEBUG   DB update was skipped because DB is the latest
2020-10-19T02:19:09.117+0100    DEBUG   DB Schema: 1, Type: 1, UpdatedAt: 2020-10-19 12:19:49.968147945 +0000 UTC, NextUpdate: 2020-10-20 00:19:49.968147445 +0000 UTC
2020-10-19T02:19:09.117+0100    DEBUG   Vulnerability type:  [os library]
2020-10-19T02:19:09.132+0100    DEBUG   Artifact ID: sha256:d05810661b660e9dde2a52ab1231a7c64d441fab50725bb4cceb39b4af9a1690
2020-10-19T02:19:09.133+0100    DEBUG   Blob IDs: [sha256:d05810661b660e9dde2a52ab1231a7c64d441fab50725bb4cceb39b4af9a1690]
2020-10-19T02:19:09.133+0100    WARN    OS is not detected and vulnerabilities in OS packages are not detected.
2020-10-19T02:19:09.133+0100    DEBUG   Detecting library vulnerabilities, path: Gemfile.lock
2020-10-19T02:19:09.133+0100    INFO    Detecting ruby vulnerabilities...

Gemfile.lock
============
Total: 1 (UNKNOWN: 0, LOW: 0, MEDIUM: 0, HIGH: 0, CRITICAL: 1)

+---------------+------------------+----------+-------------------+------------------+--------------------------------+
|    LIBRARY    | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |  FIXED VERSION   |             TITLE              |
+---------------+------------------+----------+-------------------+------------------+--------------------------------+
| activesupport | CVE-2020-8165    | CRITICAL | 6.0.3-4           | 6.0.3.1, 5.2.4.3 | rubygem-activesupport:         |
|               |                  |          |                   |                  | potentially unintended         |
|               |                  |          |                   |                  | unmarshalling of user-provided |
|               |                  |          |                   |                  | objects in MemCacheStore and   |
|               |                  |          |                   |                  | RedisCacheStore                |
+---------------+------------------+----------+-------------------+------------------+-------------------------------
```

**Additional details (base image name, container registry info...):**

Scanning the following `Gemfile.lock` (this is enough to trigger the scan, no image only this file when scanning filesystem .)
```
PATH
  remote: .
  specs:
    test (0.0.1)
      activesupport (~> 6.0.3.3)

GEM
  remote: https://rubygems.org/
  specs:
    activesupport (6.0.3.4)
      concurrent-ruby (~> 1.0, >= 1.0.2)
      i18n (>= 0.7, < 2)
      minitest (~> 5.1)
      tzinfo (~> 1.1)
      zeitwerk (~> 2.2, >= 2.2.2)

PLATFORMS
  ruby

DEPENDENCIES

RUBY VERSION
   ruby 2.7.1p83

BUNDLED WITH
   2.1.4
```

**Comment by Johannestegner** [comment]
I think this might be an issue with the parsing of the package, currently Trivy expects a version to be in semver format while the format of the specific package seems to be in the same format as legacy nuget packages. I'm currently working on that in #686.  

The big difference here seems to be that the final value does _not_ indicate build version. So I kind of wonder what it means in the ruby ecosystem compared to nuget... If we could parse it as `x.x.x+rev.x` I think it would be possible to solve this in the same way as the NuGet code will/may work.

**Comment by knqyf263** [comment]
Thank you for reporting! We're working on the issue in #740.

**Comment by knqyf263** [comment]
I developed [go-gem-version](https://github.com/aquasecurity/go-gem-version) finally. I believe #740 addressed this issue and it will be delivered at v0.13.0. If you still have any issues, please feel free to reopen the issue.
